File: packages/guards/src/numeric-range-check.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface NumericRangeCheckOptions {
  action: 'block' | 'warn';
}

const PERCENT_RE = /(\d+(?:\.\d+)?)\s*%/g;
const PROBABILITY_RE = /probability\s+(?:of\s+)?(\d+(?:\.\d+)?)/gi;
const TEMPERATURE_RE = /(-?\d+(?:\.\d+)?)\s*°\s*[CF]/g;
const NEGATIVE_COUNT_RE = /(-\d+)\s+(items?|people|users?|counts?|results?|records?)/gi;

export function numericRangeCheck(options: NumericRangeCheckOptions): Guard {
  return {
    name: 'numeric-range-check',
    version: '0.1.0',
    description: 'Validates numeric values in output',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      let m: RegExpExecArray | null;
      const pctRe = new RegExp(PERCENT_RE.source, 'g');
      while ((m = pctRe.exec(text)) !== null) {
        const v = parseFloat(m[1]);
        if (v > 100 || v < 0) {
          issues.push(`invalid_percent:${v}`);
        }
      }

      const probRe = new RegExp(PROBABILITY_RE.source, 'gi');
      while ((m = probRe.exec(text)) !== null) {
        const v = parseFloat(m[1]);
        if (v > 1 && v < 2) {
          issues.push(`invalid_probability:${v}`);
        }
      }

      const tempRe = new RegExp(TEMPERATURE_RE.source, 'g');
      while ((m = tempRe.exec(text)) !== null) {
        const v = parseFloat(m[1]);
        const unit = m[0].slice(-1);
        if (unit === 'C' && (v > 1000 || v < -273.15)) {
          issues.push(`unreasonable_temp_C:${v}`);
        }
        if (unit === 'F' && (v > 2000 || v < -459.67)) {
          issues.push(`unreasonable_temp_F:${v}`);
        }
      }

      const negRe = new RegExp(NEGATIVE_COUNT_RE.source, 'gi');
      while ((m = negRe.exec(text)) !== null) {
        issues.push(`negative_count:${m[1]}`);
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 3, 1.0) : 0;

      return {
        guardName: 'numeric-range-check',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/numeric_range_check.py
"""Validate numeric values in output."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_PERCENT_RE = re.compile(r"(\d+(?:\.\d+)?)\s*%")
_PROBABILITY_RE = re.compile(
    r"probability\s+(?:of\s+)?(\d+(?:\.\d+)?)",
    re.IGNORECASE,
)
_TEMPERATURE_RE = re.compile(
    r"(-?\d+(?:\.\d+)?)\s*\u00b0\s*([CF])"
)
_NEGATIVE_COUNT_RE = re.compile(
    r"(-\d+)\s+(items?|people|users?|counts?"
    r"|results?|records?)",
    re.IGNORECASE,
)


class _NumericRangeCheck:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "numeric-range-check"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        for m in _PERCENT_RE.finditer(text):
            v = float(m.group(1))
            if v > 100 or v < 0:
                issues.append(f"invalid_percent:{v}")

        for m in _PROBABILITY_RE.finditer(text):
            v = float(m.group(1))
            if 1 < v < 2:
                issues.append(
                    f"invalid_probability:{v}"
                )

        for m in _TEMPERATURE_RE.finditer(text):
            v = float(m.group(1))
            unit = m.group(2)
            if unit == "C" and (v > 1000 or v < -273.15):
                issues.append(
                    f"unreasonable_temp_C:{v}"
                )
            if unit == "F" and (v > 2000 or v < -459.67):
                issues.append(
                    f"unreasonable_temp_F:{v}"
                )

        for m in _NEGATIVE_COUNT_RE.finditer(text):
            issues.append(
                f"negative_count:{m.group(1)}"
            )

        triggered = len(issues) > 0
        score = min(len(issues) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="numeric-range-check",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Numeric range issue detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def numeric_range_check(
    *, action: str = "warn"
) -> _NumericRangeCheck:
    return _NumericRangeCheck(action=action)
